src/sqlfluff/core/linter.py

```diff
@@ -923,99 +923,95 @@ def extract_ignore_from_comment(comment: RawSegment):
-                        # Check for infinite loops
-                        if new_working.raw not in previous_versions:
+                        new_tree, _ = tree.apply_fixes(fixes)
+                        if new_tree.raw not in previous_versions:
                             # We've not seen this version of the file so far. Continue.
-                            working = new_working
-                            previous_versions.add(working.raw)
+                            tree = new_tree
+                            previous_versions.add(tree.raw)
                             changed = True
                             continue
-                        # Applying these fixes took us back to a state which we've
-                        # seen before. Abort.
-                        linter_logger.warning(
-                            "One fix for %s not applied, it would re-cause the same error.",
-                            crawler.code,
-                        )
-            # Keep track of initial errors for reporting.
-            if fix_loop_idx == 1:
-                initial_linting_errors = linting_errors.copy()
-            # We did not change the file. Either the file is clean (no fixes), or
-            # any fixes which are present will take us back to a previous state.
-            if not changed:
+                        else:
+                            self._warn_unfixable(crawler.code)
+
+            if loop == 0:
+                # Keep track of initial errors for reporting.
+                initial_linting_errors = all_linting_errors.copy()
+
+            if not fix:
+                # If we just want to lint then we only need one pass
+                break
+
+            if fix and not changed:
+                # We did not change the file. Either the file is clean (no fixes), or
+                # any fixes which are present will take us back to a previous state.
                 linter_logger.info(
-                    "Fix loop complete. Stability achieved after %s/%s loops.",
-                    fix_loop_idx,
-                    loop_limit,
+                    f"Fix loop complete. Stability achieved after {loop}/{loop_limit} loops."
                 )
                 break
-        else:
-            linter_logger.warning(
-                "Loop limit on fixes reached [%s]. Some fixes may be overdone.",
-                loop_limit,
-            )
-        return working, initial_linting_errors
+        if loop + 1 == loop_limit:
+            linter_logger.warning(f"Loop limit on fixes reached [{loop_limit}].")
+
+        source_only_linting_errors = self.remove_templated_errors(
+            initial_linting_errors
+        )
+
+        return tree, source_only_linting_errors
+
+    def remove_templated_errors(
+        self, initial_linting_errors: List[SQLLintError]
+    ) -> List[SQLLintError]:
+        """Filter a list of lint errors, removing those which only occur in templated slices."""
+        # TODO
```

Comment:
@alanmcruickshank Here is where I believe we need to filter out any linting errors which have been raised in templated slices. I've encountered a slight complication to this though with L003 in the added test:

raw:
```
select * from a
```

templated:
```
with dbt__CTE__INTERNAL_test as (
select * from a
)select count(*) from dbt__CTE__INTERNAL_test
```

And the lint error `Indent expected and not found compared to line #2` where line #2 is `with dbt__CTE__INTERNAL_test as (`, and the lint error is being raised on line #3, `select * from a`. I'm not sure how we'll get round this one, as the linting error is on a literal slice, but is being triggered as the result of a templated slice.